RandLAPACK/gpu_functions/rl_cuda_kernels.cuh

```diff
@@ -0,0 +1,934 @@
+    }
+}
+
+template <typename T>
+__global__  void __launch_bounds__(128) ger_gpu(
+    int64_t m, 
+    int64_t n, 
+    T alpha, 
+    T *x, 
+    int64_t incx, 
+    T *y, 
+    int64_t incy, 
+    T* A, 
+    int64_t lda
+) {
+    #pragma unroll
+    for(int i = 0; i < n; ++i) {
+        axpy_gpu_device(m, alpha * y[i * incy], x, incx, &A[i * lda], 1);
+    }
+}
+
+template <typename T>
+__device__  void ger_gpu_device(
+    int64_t m, 
+    int64_t n, 
+    T alpha, 
+    T *x, 
+    int64_t incx, 
+    T *y, 
+    int64_t incy, 
+    T* A, 
+    int64_t lda
+) {
+    #pragma unroll
+    for(int i = 0; i < n; ++i) {
+        axpy_gpu_device(m, alpha * y[i * incy], x, incx, &A[i * lda], 1);
+    }
+}
+
+template <typename T>
+__global__  void __launch_bounds__(128) get_U_gpu(
+    int64_t m,
+    int64_t n,
+    T* A,
+    int64_t lda
+) {
+    int i = blockIdx.x * blockDim.x + threadIdx.x;
+    if (i < n - 1) {
+        std::fill(&A[i * (lda + 1) + 1], &A[(i * lda) + m], 0.0);
+    }
+}
+
+/// Positions columns of A in accordance with idx vector of length k.
+/// idx array is to modified modified ONLY within the scope of this function.
+template <typename T>
+__global__ void __launch_bounds__(128) col_swap_gpu_kernel_sequential(
+    int64_t m, 
+    int64_t n, 
+    int64_t k,
+    T* A, 
+    int64_t lda,
+    int64_t const* idx,
+    int64_t* idx_mut
+) {
+    if (k > n) {
+        return;
+    }
+    cooperative_groups::grid_group g = cooperative_groups::this_grid();
+    A -= lda;
+    for (int64_t i = threadIdx.x + blockDim.x * blockIdx.x; i < k; i += blockDim.x * gridDim.x) {
+        idx_mut[i] = idx[i];
+    }
+    g.sync();
+    int64_t* curr = idx_mut;
+    int64_t* end = idx_mut + k;
+    for (int64_t i = 1; i <= k; ++i, ++curr) {
+        // swap rows IFF mismatched
+        if (int64_t const j = *curr; i != j) {
+            for (int64_t l = threadIdx.x + blockDim.x * blockIdx.x; l < m; l += blockDim.x * gridDim.x) {
+                std::iter_swap(A + i * lda + l, A + j * lda + l);
+            }
+            if (threadIdx.x == 0 && blockIdx.x == 0) {
+                std::iter_swap(curr, std::find(curr, end, i));
+            }
+            g.sync();
+        }
+    }
+}
+
+/// Positions columns of A in accordance with idx vector of length k.
+/// idx array is to modified modified ONLY within the scope of this function.
+template <typename T>
+__global__ void col_swap_gpu_kernel(
+    int64_t m, 
+    int64_t n, 
+    int64_t k,
+    T* A, 
+    int64_t lda,
+    T* A_cpy, 
+    int64_t ldac,
+    int64_t const* idx
+) {
+    /*
+    for (int64_t colIdx = threadIdx.x + blockDim.x * blockIdx.x; colIdx < k; colIdx += blockDim.x * gridDim.x) {
+        for (int64_t rowIdx = threadIdx.y + blockDim.y * blockIdx.y; rowIdx < m; rowIdx += blockDim.y * gridDim.y) {
+                int64_t j = idx[colIdx] - 1; 
+                A[colIdx * lda + rowIdx] = A_cpy[j * ldac + rowIdx];
+            }
+        }
+    }
+    */
```

Comment:
Please remove this code or explain why it should be here.

Reply:
The commented-out method should be faster, but apparently, something is off with it.